modules/mod-deep-learning/DeepLearningEffectBase.cpp

```diff
@@ -0,0 +1,363 @@
+         
+         blockIndices.emplace_back(BlockIndex(samplePos, blockSize));
+
+         samplePos += blockSize;
+      }
+   }
+
+   return blockIndices;
+}
+
+std::vector<ClipTimestamps> DeepLearningEffectBase::GetClipTimestamps(WaveTrack *track, double tStart, double tEnd) const
+{
+   std::vector<ClipTimestamps> timestamps;
+
+   for (const auto &clip : track->GetClips())
+      timestamps.emplace_back(
+         std::pair<double, double>(clip->GetPlayStartTime(), 
+                                   clip->GetPlayEndTime())
+      );
+
+   return timestamps;
+}
+
+
+torch::Tensor DeepLearningEffectBase::BuildMonoTensor(WaveTrack *track, float *buffer,
+                                                  sampleCount start, size_t len)
+{
+   //Get the samples from the track and put them in the buffer
+   if (!track->GetFloats(buffer, start, len))
+      throw std::runtime_error("An error occurred while copying samples to tensor buffer.");
+
+   // get tensor input from buffer
+   torch::Tensor audio = torch::from_blob(buffer, len,
+                                          torch::TensorOptions().dtype(torch::kFloat));
+   audio = audio.unsqueeze(0); // add channel dimension
+
+   return audio;
+}
+
+torch::Tensor DeepLearningEffectBase::BuildMultichannelTensor(WaveTrack *leader, float *buffer,
+                                                          sampleCount start, size_t len)
+{
+   auto channels = TrackList::Channels(leader);
+   std::vector<torch::Tensor> channelStack;
+
+   // because we're reusing the same buffer, it's important that
+   // we clone the tensor.
+   for (WaveTrack *channel : channels)
+      channelStack.emplace_back(
+          BuildMonoTensor(channel, buffer, start, len).clone());
+
+   return torch::cat(channelStack);
+}
+
+torch::jit::IValue DeepLearningEffectBase::ForwardPassInThread(torch::Tensor input)
+{
+   torch::jit::IValue output;
+
+   std::atomic<bool> done = {false};
+   std::atomic<bool> success = {true};
+
+   // make a copy of the model (in case we need to abort)
+   DeepModelHolder model = this->mModel;
+
+   auto thread = std::thread(
+      [model, &input, &output, &done, &success]()
+      {
+         try
+         {
+            torch::jit::IValue tempOut = model->Forward(input);
+
+            // only write to output tensor if abort was not requested
+            if (success)
+               output = tempOut;
+         }
+         catch (const ModelException &e)
+         {
+            wxLogError(e.what());
+            wxLogDebug(e.what());
+            success = false;
+            output = torch::jit::IValue(torch::zeros_like(input));
+         }
+         done = true;
+      }
+   );
+
+   // wait for the thread to finish
+   while (!done)
+   {
+      if (TrackProgress(mCurrentTrackNum, mCurrentProgress))
+      {
+         // abort if requested
+         success = false;
+
+         // tensor output will be destroyed once the thread is destroyed
+         thread.detach();
```

Comment:
In the case that the thread detaches, because the user clicks the cancel button in progress, are we sure it will exit before too much time and not keep consuming resources?  There is no portable way to kill a std::thread that invokes bad code.

Reply:
There is a greater danger in the case of a detached thread.  The lambda capture of the thread function will have dangling references to stack variables of this function after it exits!

There is also still a data race updating `output` in this path.  The thread might be in the middle of assignment to output before it sees the change of state of `success`.

I think you should to this:

Make a shared_ptr to a structure containing variables that the thread needs.  Capture that shared_ptr BY VALUE in the thread function.  If you take the path that joins the thread, copy the `torch::jit::IValue` that the thread wrote into a result variable.  (Copies of that type are inexpensive?) But if you take the path that detaches the thread, then zero out the result instead; let the detached thread keep running, but never touching the result variable.

Reply:
done! added a struct that holds all necessary variables for as long as the detached thread runs